### DISCUSSION TITLE
Override Dependencies for an endpoint only in the Docs

### DISCUSSION THREAD

**Discussion by Hyperclaw79** [fastapi/fastapi#10137] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typing import Annotated
from fastapi import FastAPI, Depends, Request
from fastapi.security import OAuth2PasswordBearer


app = FastAPI(
    title="StocksALot API",
    summary="External API to access the stocksalot data and insights."
)


async def get_current_user(token: Annotated[
    str, Depends(OAuth2PasswordBearer(tokenUrl="token"))
]) -> str:
    """Get the current user."""
    # Auth logic
    return "external dude"


async def get_username(request: Request):
    """Check if the request is internal."""
    internal_client = request.headers.get("X-Internal-Client")
    k8s_token = request.headers.get("X-Internal-Token")
    if (
        internal_client and k8s_token
        and k8s_token == "test"
    ):
        return "internal"
    token = await OAuth2PasswordBearer(tokenUrl="token")(request)
    return await get_current_user(token)


@app.get('/ohlc')
async def get_ohlc(username: Annotated[
    str, Depends(get_username)
]):
    """Get all OHLC data."""
    if username != "internal":
        print("User %s requested all OHLC data.", username)
    # Rest of the logic

if __name__ == "__main__":
    import uvicorn
    uvicorn.run(app, host="0.0.0.0", port=5000)
```

### Description

I have a use case where I am getting traffic from external users and another microservice inside the same kubernetes cluster. Due to this I want to reuse the endpoints for both the types of traffic but segregate them using conditional dependency injection.
Everything works fine for the API. However, on the `/docs`,  since `get_ohlc` now uses the dependency injection of `get_username` instead of `get_current_user`, the param is no longer is authentication locked.

Is there a way to override the behavior exclusively for the docs?\
If not, I'm open to design change suggestions as well as long as the core functionality remains intact.

Please help me in getting this sorted.

### Operating System

Linux, Other

### Operating System Details

I'm running it inside a Kubernetes Cluster using the latest Python Alpine as the base image.
K8s version - v1.27.2

### FastAPI Version

0.101.1

### Pydantic Version

2.2.1

### Python Version

Python 3.11.1

### Locked by OAuth Password form when using `get_current_user` as the dependency:
<img width="1124" alt="image" src="https://github.com/tiangolo/fastapi/assets/29298411/daf81a99-93c7-4c81-ba78-8124d01e756b">

### Authentication Error but no way to let users know that Oauth token is required:
<img width="1125" alt="image" src="https://github.com/tiangolo/fastapi/assets/29298411/e2926c43-388a-4aaa-8a7b-16406ddf72e8">

**Comment by Hyperclaw79** [comment, marked as answer]
Came up with a workaround by using parameterized dependency and `auto_error=False` in Oauth2PasswordBearer.
However, this is a partial fix since even though the user can see the requirement for token, there is no more autolocking the request form.

**Comment by chrisK824** [comment]
Maybe I'm losing something here, but why can you make both things, authentication and username dissection based on headers, in the same function as a dependency? You could pass the Request object as parameter and pass the Oauth bearer token dependency as well in the same function. First retrieve the headers to determine source of request, then based on condition either do your "internal" handle or proceed with normal authentication and get current user functionality.
